contracts/BalanceTrackerFixedPriceBase.sol

```diff
@@ -0,0 +1,242 @@
+/// @param provided Provided payload length.
+/// @param expected Expected payload length.
+error InvalidPayloadLength(uint256 provided, uint256 expected);
+
+/// @dev Failure of a transfer.
+/// @param token Address of a token.
+/// @param from Address `from`.
+/// @param to Address `to`.
+/// @param amount Amount value.
+error TransferFailed(address token, address from, address to, uint256 amount);
+
+abstract contract BalanceTrackerFixedPriceBase {
+    event MechPaymentCalculated(address indexed mech, uint256 indexed requestId, uint256 deliveryRate, uint256 rateDiff);
+    event Deposit(address indexed account, address indexed token, uint256 amount);
+    event Withdraw(address indexed account, address indexed token, uint256 amount);
+    event Drained(address indexed token, uint256 collectedFees);
+
+    // Fee base constant
+    uint256 public constant FEE_BASE = 10_000;
+
+    // Mech marketplace address
+    address public immutable mechMarketplace;
+    // Buy back burner address
+    address public immutable buyBackBurner;
+    // Collected fees
+    uint256 public collectedFees;
+    // Reentrancy lock
+    uint256 internal _locked = 1;
+
+    // Map of requester => current debit balance
+    mapping(address => uint256) public mapRequesterBalances;
+    // Map of mech => => current debit balance
+    mapping(address => uint256) public mapMechBalances;
+
+    /// @dev BalanceTrackerFixedPrice constructor.
+    /// @param _mechMarketplace Mech marketplace address.
+    /// @param _buyBackBurner Buy back burner address.
+    constructor(address _mechMarketplace, address _buyBackBurner) {
+        // Check for zero address
+        if (_mechMarketplace == address(0) || _buyBackBurner == address(0)) {
+            revert ZeroAddress();
+        }
+
+        mechMarketplace = _mechMarketplace;
+        buyBackBurner = _buyBackBurner;
+    }
+
+    function _checkNativeValue() internal virtual;
+
+    function _getRequiredFunds(address requester, uint256 balanceDiff) internal virtual returns (uint256);
+
+    // Check and record delivery rate
+    function checkAndRecordDeliveryRate(
+        address mech,
+        address requester,
+        uint256
+    ) external payable {
+        // Check for marketplace access
+        if (msg.sender != mechMarketplace) {
+            revert MarketplaceOnly(msg.sender, mechMarketplace);
+        }
+
+        // Check for native value
+        _checkNativeValue();
+
+        // Get mech max delivery rate
+        uint256 maxDeliveryRate = IMech(mech).maxDeliveryRate();
+
+        // Get account balance
+        uint256 balance = mapRequesterBalances[requester];
+
+        // Check the request delivery rate for a fixed price
+        if (balance < maxDeliveryRate) {
+            // Get balance difference
+            uint256 balanceDiff = maxDeliveryRate - balance;
+            // Adjust balance
+            balance += _getRequiredFunds(requester, balanceDiff);
+        }
```

Comment:
This is key - the balance can be top-upped via receive / deposit function. However if there's a lack of it - charge "at place", that makes the behavior exact equivalent to what there currently is.